fcrepo4/fcrepo4
# implemented the listing of resource versions

**JIRA Ticket**: [FCREPO-3237](https://jira.lyrasis.org/browse/FCREPO-3237)

# What does this Pull Request do?
Allows resource versions to be listed.

# How should this be tested?

Note that `fcrepo.autoversioning.enabled` must be enabled.

```
mvn -Dfcrepo.autoversioning.enabled=true jetty:run

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-ag-2 -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -H'Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG Parent'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-ag-2/child1 -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG Child 1'"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-ag-2/child2 -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "testing"

curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test-ag-2/child2 -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "updated"

curl -u fedoraAdmin:fedoraAdmin -v -H "Accept: text/turtle" http://localhost:8080/rest/test-ag-2/fcr:versions

@prefix memento:  <http://mementoweb.org/ns#> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix fedora:  <http://fedora.info/definitions/v4/repository#> .
@prefix ldp:  <http://www.w3.org/ns/ldp#> .

<info:fedora/test-ag-2/fcr:versions>
        ldp:contains         <info:fedora/test-ag-2/fcr:versions/20200226203011> ;
        ldp:contains         <info:fedora/test-ag-2/fcr:versions/20200226202947> ;
        ldp:contains         <info:fedora/test-ag-2/fcr:versions/20200226202936> ;
        ldp:contains         <info:fedora/test-ag-2/fcr:versions/20200226202926> ;
        memento:original     <info:fedora/test-ag-2> ;
        fedora:created       "2020-02-26T20:29:26.774214Z" ;
        fedora:lastModified  "2020-02-26T20:29:26.774214Z" ;
        rdf:type             fedora:TimeMap ;
        rdf:type             memento:TimeMap .

curl -u fedoraAdmin:fedoraAdmin -v -H "Accept: application/link-format" http://localhost:8080/rest/test-ag-2/fcr:versions

<http://localhost:8080/rest/test-ag-2>; rel="original",
<http://localhost:8080/rest/test-ag-2>; rel="timegate",
<http://localhost:8080/rest/test-ag-2/fcr:versions/20200226202926>; rel="memento"; datetime="Wed, 26 Feb 2020 20:29:26 GMT",
<http://localhost:8080/rest/test-ag-2/fcr:versions/20200226202936>; rel="memento"; datetime="Wed, 26 Feb 2020 20:29:36 GMT",
<http://localhost:8080/rest/test-ag-2/fcr:versions/20200226202947>; rel="memento"; datetime="Wed, 26 Feb 2020 20:29:47 GMT",
<http://localhost:8080/rest/test-ag-2/fcr:versions/20200226203011>; rel="memento"; datetime="Wed, 26 Feb 2020 20:30:11 GMT",
<http://localhost:8080/rest/test-ag-2/fcr:versions>; rel="self"; from="Wed, 26 Feb 2020 20:29:26 GMT"; until="Wed, 26 Feb 2020 20:30:11 GMT"; type="application/link-format",

curl -u fedoraAdmin:fedoraAdmin -v -H "Accept: text/turtle" http://localhost:8080/rest/test-ag-2/child1/fcr:versions

@prefix memento:  <http://mementoweb.org/ns#> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix fedora:  <http://fedora.info/definitions/v4/repository#> .
@prefix ldp:  <http://www.w3.org/ns/ldp#> .

<info:fedora/test-ag-2/child1/fcr:versions>
        ldp:contains         <info:fedora/test-ag-2/child1/fcr:versions/20200226202936> ;
        memento:original     <info:fedora/test-ag-2/child1> ;
        fedora:created       "2020-02-26T20:29:36.544215Z" ;
        fedora:lastModified  "2020-02-26T20:29:36.544215Z" ;
        rdf:type             fedora:TimeMap ;
        rdf:type             memento:TimeMap .

curl -u fedoraAdmin:fedoraAdmin -v -H "Accept: text/turtle" http://localhost:8080/rest/test-ag-2/child2/fcr:versions

@prefix memento:  <http://mementoweb.org/ns#> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix fedora:  <http://fedora.info/definitions/v4/repository#> .
@prefix ldp:  <http://www.w3.org/ns/ldp#> .

<info:fedora/test-ag-2/child2/fcr:versions>
        ldp:contains         <info:fedora/test-ag-2/child2/fcr:versions/20200226203011> ;
        ldp:contains         <info:fedora/test-ag-2/child2/fcr:versions/20200226202947> ;
        memento:original     <info:fedora/test-ag-2/child2> ;
        fedora:created       "2020-02-26T20:29:47.229216Z" ;
        fedora:lastModified  "2020-02-26T20:30:11.037218Z" ;
        rdf:type             fedora:TimeMap ;
        rdf:type             memento:TimeMap .
```

# Additional Notes:

* Resource URIs are incorrect, except when `application/link-format` is used. This PR does not solve a general problem that exists mapping between different types of IDs.
* The integration tests are NOT reenabled. They still don't work because 1) the URIs are wrong and 2) you can't create versions on demand yet (just through auto-versioning).

# Interested parties
@dbernstein @awoods @whikloj
